src/main/scala/com/fulcrumgenomics/umi/ConsensusCallingIterator.scala

```diff
@@ -26,62 +26,95 @@ package com.fulcrumgenomics.umi
   */
-class ConsensusCallingIterator(sourceIterator: Iterator[SamRecord],
-                               val caller: UmiConsensusCaller[_],
-                               val progress: Option[ProgressLogger] = None
-                              ) extends Iterator[SamRecord] {
+class ConsensusCallingIterator[C <: SimpleRead](sourceIterator: Iterator[SamRecord],
+                               caller: UmiConsensusCaller[C],
+                               progress: Option[ProgressLogger] = None,
+                               threads: Int = 1,
+                               maxRecordsInRam: Int = 128000)
+  extends Iterator[SamRecord] with LazyLogging {
+  protected val groupingIterator: Iterator[Seq[SamRecord]] = {
+    new SamRecordGroupedIterator(sourceIterator, caller.sourceMoleculeId)
+      .map { records => for (p <- progress; r <- records) p.record(r); records }
+  }
+  protected val iterator: Iterator[SamRecord] = {
+    if (threads <= 1) {
+      groupingIterator.flatMap(caller.consensusReadsFromSamRecords)
+    }
+    else {
+      val pool          = new ForkJoinPool(threads, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, true)
+      val bufferedIter  = groupingIterator.bufferBetter
+      val callersFactor = 1000 // the number of callers per thread to create. This to reduce chance of picking the same caller
+      val callers       = Array.range(start=0, threads*callersFactor).map { _ => caller.emptyClone() }
+
+      // Read in groups of records (each from the same source molecule) until we have the maximum number of
+      // individual records in RAM.  We have a few more records in RAM, if the group that pushes us over the limit is
+      // large.  Then process the collected groups in parallel.
+      Iterator.continually {
+        var total = 0L
+        bufferedIter
+          .takeWhile { chunk => if (maxRecordsInRam <= total) false else {total += chunk.length; true } }
+          .zipWithIndex
+          .toSeq
+          .parWith(pool)
+          .flatMap { case (records, groupIndex) =>
+            val caller = callers(groupIndex % callers.length)
+            caller.synchronized { caller.consensusReadsFromSamRecords(records) }
+          }
+          .seq
+      }.takeWhile { records =>
+        if (records.nonEmpty) true else {
+          // add the statistics to the original caller since there are no more reads
+          require(bufferedIter.isEmpty, "Bug: input is not empty")
+          callers.foreach(caller.addStatistics)
+          false
+        }
+      }.flatten
+
+    }
+  }
+  override def hasNext: Boolean = this.iterator.hasNext
+  override def next(): SamRecord = this.iterator.next
+}
 
-  private val input = sourceIterator.bufferBetter
-  private val outputQueue: mutable.Queue[SamRecord] = mutable.Queue[SamRecord]()
+/** Groups consecutive records based on a method to group records. */
+private class SamRecordGroupedIterator[Key](sourceIterator: Iterator[SamRecord],
```

Comment:
Ugh, I didn't realize `Iterator.groupBy` doesn't exist :(  I wonder if we'd be better served adding a `groupBy` to `BetterBufferedIterator` and making this more generic?